### PR TITLE
 Stateful Models Documentation Update

### PR DESCRIPTION
**Pull request by itikhono**
Details:

- updated to OpenVINO 2.0
- added MakeStateful documentation

Tickets:
CVS-68160

### CODE REVIEW

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    :hidden:
+
+    openvino_docs_OV_UG_ways_to_get_stateful_model
+
+What is Stateful Model?
+#######################
+
+Stateful model is a model which stores the specified data between and during model inferences in internal memory space usually called State or Variable.
+In contrast to usual **stateless** model, which purify all the calculated data after each inference call, **stateful**
+model preserve tensors saved in States.
```
Comment (slyalin) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst:17]:
```suggestion
Stateful model is a model which implicitly keeps data from one inference call to the next inference call. Data is kept in internal runtime memory space usually called State or Variable.
In contrast to usual **stateless** model, which return all produced data as model outputs, **stateful**
model preserve part of the tensors saved in States without exposing them as model outputs.

The purpose of stateful models is to natively address a sequence processing like in text generation tasks when one model inference produce a single output token, and it is required to perform multiple inference calls to generate a complete output sentence. Hidden state data from previous inference should be passed to the next inference as a context. Usually the contextual data is not required to be accessed in the user application and should be just passed through to the next inference call manually using model API. Stateful models simplifies programming of this scenario and unlocks additional performance potential of OpenVINO runtime.
```

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+OpenVINO Stateful Model Representation
+######################################
+
+OpenVINO contains ReadValue/Assign operations to make a model Stateful.
+Each pair of ReadValue/Assign operates with State, known also as Variable,
+which is an internal memory buffer to store tensor data during and between model inference calls.
+ReadValue reads data tensor from State and returns it as output, Assign accepts data tensor as input and writes data to State
+to save data for the next inference call.
```
Comment (slyalin) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst:26]: Move the first paragraph to a more low-level explanation below. Rename this section to Benefits of Stateful Models and list those 3 items below without mentioning ReadValue and Assign ops. Introduce low-level stuff gradually later. The beginning of the article shouldn't touch op-level specifics, first explain purpose, benefits and ways to obtain stateful models, then move to API specifics and inherent limitations that helps runtime to unlock that benefits that were promised in the beginning in comparison to regular Parameter/Results.

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+These entities work as a coherent mechanism and help to achieve the next goals:
+
+1. Speed up execution of Model
+    Data in State is stored in the optimized form for OpenVINO plugins, which helps to execute model effectively.
+    But often requesting data from State might reduce the expected performance gains and even lead to losses,
+    so, it is expected that State mechanism will be used when data stored in State is not accessed frequently.
```
Comment (slyalin) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst:33]: We need to be more careful with performance promises. It really depends on the model architecture (patterns used between ReadValue and Assign ops in the graph). Please be more specific and state that performance improvements are expected specifically for large language models produced by Optimum-Intel path and for other models it may vary.

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+3. Specific scenarios
+    Several use cases require processing of data sequences. When length of a sequence is known and small enough,
+    we can process it with RNN like models that contain a cycle inside. But in some cases, like online speech recognition of time series
+    forecasting, length of data sequence is unknown. Then data can be divided in small portions and processed step-by-step. But dependency
+    between data portions should be addressed. For that, models save some data between inferences - state. When one dependent sequence is over,
+    state should be reset to initial value and new sequence can be started.
+
+OpenVINO has a special API to simplify work with Stateful models. State is automatically saved between inferences,
+and there is a way to reset state when needed. You can also read state or set it to some new value between inferences.
```
Comment (slyalin) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst:47]: Please rephrase it to be more fluent: we have already state that we have a dedicated API for that. And it looks to be a duplication. Provide more details (e.g. InferRequest class reference to identify where this API exists).

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+#########################################
+
+* [Apply LowLatency2 transformation.](./ways_to_get_stateful_model.md#)
+   If a model contains a loop that runs over some sequence of input data,
+   the LowLatency2 transformation can be applied to get model with states.
+   Note: there are some [specific limitations]() to use the transformation.
+
+* [Apply MakeStateful transformation.](./ways_to_get_stateful_model.md)
+   If after conversion from original model to OpenVINO representation, the resulting model contains Parameter and Result operations,
+   which pairwise have the same shape and element type, the MakeStateful transformation can be applied to get model with states.
```
Comment (slyalin) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst:80]: Promote Optimum-Intel as the main source of stateful models. Refer to Gen AI article in this documentation. Position LowLatency2 and MakeStateful as more low-level/advanced technique for developers who really know what they want.

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    But often requesting data from State might reduce the expected performance gains and even lead to losses,
+    so, it is expected that State mechanism will be used when data stored in State is not accessed frequently.
+
+2. Simplify user code
+    Typical scenarios as copying data from model's outputs to inputs in user code can be replaced with ReadValue/Assign operations.
+    OpenVINO will manage this copying internally.
+
+3. Specific scenarios
+    Several use cases require processing of data sequences. When length of a sequence is known and small enough,
+    we can process it with RNN like models that contain a cycle inside. But in some cases, like online speech recognition of time series
```
Comment (Wovchena) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst:41]:
```suggestion
    we can process it with RNN like models that contain a cycle inside. But in some cases, like online speech recognition or time series
```

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst (hunk 3 above)
Comment (slyalin) [reply docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst:33]: I mean even state is not accessed and developer plays by the rules, it may impact performance for some architectures. It is hard to explain when it happens, it is better to point to text generation with model produced by Optimum-Intel where we are definitely expect performance gain.

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+and there is a way to reset state when needed. You can also read state or set it to some new value between inferences.
+
+Stateful Model Example
+######################
+
+
+.. image:: _static/images/stateful_model_example.png
+   :align: center
+
+The left side of the picture shows the usual inputs and outputs to the model: Parameter/Result operations. They have no connection with each other and in order to copy data from output to input users need to put extra effort writing and maintaining additional code.
```
Comment (Wovchena) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst:56]:
```suggestion
The left side of the picture shows the usual inputs and outputs to the model: Parameter/Result operations. There's no direct connection from Result to Parameter and in order to copy data from output to input users need to put extra effort writing and maintaining additional code.
```
In fact there are always two connections, just indirect: Parameter->Result through model and Result->Parameter through code

File: docs/sphinx_setup/_static/images/make_stateful_simple.png
Comment (ilya-lavrenov) [inline docs/sphinx_setup/_static/images/make_stateful_simple.png:1]: please, put under LFS

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+By default, the LowLatency2 transformation inserts a constant subgraph of the same shape as the previous input node, and with zero values as the initializing value for ReadValue nodes, please see the picture below. We can disable insertion of this subgraph by passing the `false` value for the `use_const_initializer` argument.
+
+.. tab:: C++
+
+      .. doxygensnippet:: docs/snippets/ov_model_state_intro.cpp
+         :language: cpp
+         :fragment: [ov:low_latency_2_use_parameters]
+
+
+.. image:: _static/images/llt2_use_const_initializer.png
```
Comment (tsavina) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst:84]:
```suggestion
.. image:: _static/images/llt2_use_const_initializer.svg
```

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+LowLatencу2
+###########
+
+LowLatency2 transformation changes the structure of the model containing [TensorIterator](../ops/infrastructure/TensorIterator_1.md) 
+and [Loop](../ops/infrastructure/Loop_5.md) by adding the ability to work with the state, inserting the Assign/ReadValue 
+layers as it is shown in the picture below.
+
+Example of applying LowLatency2 transformation:
+
+.. image:: _static/images/applying_low_latency_2.png
```
Comment (tsavina) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst:34]:
```suggestion
.. image:: _static/images/applying_low_latency_2.svg
```

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+
+4. Use state API. See sections [OpenVINO state API](#openvino-state-api), [Example of stateful model inference](#example-of-stateful-model-inference).
+
+### Known Limitations
+1. Unable to execute [Reshape](ShapeInference.md) to change the number iterations of TensorIterator/Loop layers to apply the transformation correctly due to hardcoded values of shapes somewhere in the model.
+
+   The only way you can change the number iterations of TensorIterator/Loop layer is to use the Reshape feature, but models can be non-reshapable, the most common reason is that the value of shapes is hardcoded in a constant somewhere in the model.
+
+.. image:: _static/images/low_latency_limitation_2.png
```
Comment (tsavina) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst:102]:
```suggestion
.. image:: _static/images/low_latency_limitation_2.svg
```

File: docs/articles_en/openvino_workflow/running_inference_with_openvino.rst [hunk 11]
```diff
@@ -13,7 +13,7 @@ Running Inference with OpenVINO™
    openvino_docs_OV_UG_Working_with_devices
```
Comment (tsavina) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino.rst:13]: build error indicates that openvino_docs_OV_UG_Working_with_devices and openvino_docs_ov_ug_supported_plugins_gna contain a link to deleted article openvino_docs_OV_UG_model_state_intro, please delete all references to this article

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+Only strict syntax is supported, as in the example above, the transformation call must be in double quotes
+"MakeStateful[...]", the tensor names in single quotes 'tensor_name_1' and without spaces.
+
+## How to create a model with state using OpenVINO
+
+To get a model with states ready for inference, you can convert a model from another framework to IR with Model Optimizer 
+or create an OpenVINO Model (details can be found in :doc:`Build OpenVINO Model section<openvino_docs_OV_UG_Model_Representation>`.
+Let's build the following graph using C++ OpenVINO API:
+
+.. image:: _static/images/stateful_model_example.png
```
Comment (tsavina) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst:170]:
```suggestion
.. image:: _static/images/stateful_model_example.svg
```

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+Example of applying LowLatency2 transformation:
+
+.. image:: _static/images/applying_low_latency_2.svg
+
+After applying the transformation, ReadValue operations can receive other operations as an input, as shown in the picture above. 
+These inputs should set the initial value for initialization of ReadValue operations. 
+However, such initialization is not supported in the current State API implementation. 
+Input values are ignored and the initial values for the ReadValue operations are set to zeros unless otherwise specified 
+by the user via :doc:`State API<ov_ug_state_api>`.
```
Comment (tsavina) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst:42]:
```suggestion
by the user via :ref:`State API <ov_ug_state_api>`.
```

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst (hunk 5 above)
Comment (itikhono) [reply docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro.rst]: done

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+**State naming rule:**  a name of a state is a concatenation of names: original TensorIterator operation, Parameter of the body, and additional suffix "variable_" + id (0-base indexing, new indexing for each TensorIterator). You can use these rules to predict what the name of the inserted State will be after the transformation is applied. For example:
+
+.. tab:: C++
+
+      .. doxygensnippet:: docs/snippets/ov_model_state_intro.cpp
+         :language: cpp
+         :fragment: [ov:low_latency_2]
+
+
+4. Use state API. See sections :doc:`OpenVINO State API<ov_ug_state_api>`, [Example of stateful model inference](#example-of-stateful-model-inference).
```
Comment (tsavina) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst:97]:
```suggestion
4. Use state API. See sections :ref:`OpenVINO State API <ov_ug_state_api>`, [Example of stateful model inference](#example-of-stateful-model-inference).
```

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+.. image:: _static/images/low_latency_limitation_2.svg
+
+   **Current solution:** Trim non-reshapable layers via `ModelOptimizer CLI<openvino_docs_MO_DG_prepare_model_convert_model_Converting_Model> `--input`, `--output`. For example, the parameter and the problematic constant in the picture above can be trimmed using the following command line option:
+   `--input Reshape_layer_name`. The problematic constant can be also replaced using OpenVINO, as shown in the example below.
+
+.. tab:: C++
+
+      .. doxygensnippet:: docs/snippets/ov_model_state_intro.cpp
+         :language: cpp
+         :fragment: [ov:replace_const]
```
Comment (kblaszczak-intel) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst:114]:
```suggestion
   .. tab:: C++

      .. doxygensnippet:: docs/snippets/ov_model_state_intro.cpp
         :language: cpp
         :fragment: [ov:replace_const]
```

**Review by kblaszczak-intel — APPROVED** [review]
The change requires some polishing still. We think it would be beneficial to the user to merge it already and publish in 23.3, so the content is available, even if formatting or language issues are visible. 
When it is merged, we'll make sure to create a PR with necessary changes and suggestions, as well as port the whole thing to master.

File: docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+The only way you can change the number iterations of TensorIterator/Loop layer is to use the Reshape feature, but models can be non-reshapable,
+the most common reason is that the value of shapes is hardcoded in a constant somewhere in the model.
+
+
+.. image:: _static/images/low_latency_limitation_2.svg
+   :scale: 70 %
+   :align: center
+
+**Solution:**
+Trim non-reshapable layers via :doc:`ModelOptimizer CLI<openvino_docs_MO_DG_prepare_model_convert_model_Converting_Model> `--input`, `--output`.
```
Comment (kblaszczak-intel) [inline docs/articles_en/openvino_workflow/running_inference_with_openvino/stateful_models_intro/ways_to_get_stateful_model.rst:115]:
```suggestion
Trim non-reshapable layers via :doc:`ModelOptimizer CLI <openvino_docs_MO_DG_prepare_model_convert_model_Converting_Model>` `--input`, `--output`.
```

**Review by slyalin — APPROVED** [review]
I agree to merge the current version as a starting point. Please merge and improve in the next PR(s).
